### PR TITLE
Implement validation checks for failuredomain fields

### PR DESCRIPTION
**Pull request by charles-large**
*Issue #, if available:*
[3004](https://github.com/aws/eks-anywhere-internal/issues/3004)
*Description of changes:*
Broke up ValidateVCenterSetupMachineConfig method into three sub functions: 
 - getDatastorePath 
 - getFolderPath
 - getResourcePoolPath

New ValidateFailureDomainConfig validation logic to ensure existence of fields in vcenter

*Testing (if applicable):*
New tests in validator_test.go
*Documentation added/planned (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by charles-large** [comment]
I have two more commits based on the changes here. Pasting them here for transparency.

[Add failure domain compute cluster validation](https://github.com/aws/eks-anywhere/commit/9f5adf864538bd8107475506b084ec4de691ee9f)

[Integrate failure domain fields into missing permissions logic](https://github.com/aws/eks-anywhere/commit/074fe414c788e6b416a4d7c7944fbdc033b9d827)

File: pkg/providers/vsphere/validator.go [hunk 1]
```diff
@@ -78,47 +78,81 @@ func (v *Validator) ValidateVCenterConfig(ctx context.Context, datacenterConfig
...
 
-	for _, wng := range vsphereClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+	failureDomainsAssigned, err := v.validateWorkerNodeGroupDomains(vsphereClusterSpec, providedFailureDomains)
+	if err != nil {
+		return err
+	}
+
+	if !failureDomainsAssigned {
+		// TODO: Error message here if Failure Domain not being used by workernodegroups?
+		// Skipping further validation currently
```
Comment (vivek-koppuru) [inline pkg/providers/vsphere/validator.go:111]: Will this be a later validation?

File: pkg/executables/govc.go [hunk 2]
```diff
@@ -756,23 +756,22 @@ func (g *Govc) NetworkExists(ctx context.Context, network string) (bool, error)
 	return exists, nil
 }
 
-func (g *Govc) ValidateVCenterSetupMachineConfig(ctx context.Context, datacenterConfig *v1alpha1.VSphereDatacenterConfig, machineConfig *v1alpha1.VSphereMachineConfig, _ *bool) error {
-	envMap, err := g.validateAndSetupCreds()
+// GetDatastorePath validates and returns the full path to a datastore in the specified datacenter.
+// Returns an error if the datastore doesn't exist or if the path is invalid.
+func (g *Govc) GetDatastorePath(ctx context.Context, datacenter string, datastorePath string, envMap map[string]string) (string, error) {
```
Comment (vivek-koppuru) [inline pkg/executables/govc.go:761]: Any reason why we are passing in envMap now instead of using g.envMap or something? Just curious not saying we do that.

**Review by vivek-koppuru — COMMENTED** [review]
Wow I love the work put in here to better refactor!

Did we get a chance to do functional testing here?

File: pkg/executables/govc.go (hunk 2 above)
Comment (charles-large) [reply pkg/executables/govc.go:761]: In ValidateVCenterSetupMachineConfig and ValidateFailureDomainConfig, g.validateAndSetupCreds is executed once to setup the environment, and passed as a parameter to the GetPath methods. To avoid needing to execute it multiple times.

[Ex](https://github.com/aws/eks-anywhere/pull/9466/files/1c8a7e8dec3e415f0dfa8767d2008ca5d5e4d2fe#diff-1b2bffa403bbd82a1ba98650cb15df32447b074ad23e1cc1cf7df592eca4554dR876)

File: pkg/providers/vsphere/validator.go (hunk 1 above)
Comment (charles-large) [reply pkg/providers/vsphere/validator.go:111]: It depends on how we want to handle it. If a user in their cluster spec file defines failure domains on their datacenter, but does not assign them to any workernodegroups, than this would trigger. Since no workernodegroup is using a failuredomain, we return here and skip further validations since the failure domain feature is not being used. We could error here saying "hey you defined failure domains in your datacenter, but did not assign them to any workernodegroups" or leave the current behavior where it simply skips/ignores.

File: pkg/executables/govc.go (hunk 2 above)
Comment (charles-large) [reply pkg/executables/govc.go:761]: This could be moved back into each GetPath method to avoid needing to pass it as a parameter, but this would increase the number of times the function is executed from 1->3. It's an idempotent function though I did it for sake of efficiency.

**Comment by charles-large** [comment]
> Wow I love the work put in here to better refactor!
> 
> Did we get a chance to do functional testing here?

I considered adding unit tests for these methods but decided against it since they are primarily simple orchestration functions that call existing GetPath methods. The GetPath methods themselves I created extensive unit test coverage, and adding tests here would largely duplicate that existing coverage at a higher level. Success cases are already verified through existing test functions. If these methods become more complex in the future, I think that would be a better time to implement them. I'm open to doing them, this was just my thought process.

File: pkg/executables/govc.go [hunk 3]
```diff
@@ -781,79 +780,151 @@ func (g *Govc) ValidateVCenterSetupMachineConfig(ctx context.Context, datacenter
...
+		return err
+	}
+	machineConfig.Spec.ResourcePool = resourcePool
+
+	return nil
+}
+
+// ValidateFailureDomainConfig validates that all resources specified in a VSphere
+// failure domain exist and are accessible.
+func (g *Govc) ValidateFailureDomainConfig(ctx context.Context, datacenterConfig *v1alpha1.VSphereDatacenterConfig, failureDomain *v1alpha1.FailureDomain) error {
```
Comment (rajeshvenkata) [inline pkg/executables/govc.go:904]: Are we also validating computeCluster field?

File: pkg/providers/vsphere/validator.go (hunk 1 above)
Comment (rajeshvenkata) [reply pkg/providers/vsphere/validator.go:111]: Got it. Is there any other change needed here other than returning error? If it is just returning error, can we include this now?

File: pkg/executables/govc.go (hunk 3 above)
Comment (charles-large) [reply pkg/executables/govc.go:904]: this would be in my next commit. I was awaiting approval here, and felt breaking it into multiple commits would make it easier to review. See my comments [here](https://github.com/aws/eks-anywhere/pull/9466#issuecomment-2759500489).

File: pkg/providers/vsphere/validator.go (hunk 1 above)
Comment (charles-large) [reply pkg/providers/vsphere/validator.go:111]: changing it to an error would break some existing failure domain tests elsewhere. I didn't get a change to dive much deeper so I left it as is upon waiting review. I could look into this more.

**Comment by charles-large** [comment]
> Did we get a chance to test these changes by creating a cluster with failureDomain setup?

yes, I have validated the creation works as expected. However, let me document.

**Comment by charles-large** [comment]
Cluster is being created successfully with this change. Worker node groups in separate failure domains.

image
image
